File: 03-stream-processing/solution-proposal/python/exercise2/streaming_job.py
from pyspark.sql import SparkSession

from exercise2.configuration import get_sink_configuration
from exercise2.streaming_job_processing_logic import process_input_data

spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0') \
    .getOrCreate()

input_data_stream = spark.readStream \
    .option("kafka.bootstrap.servers", "localhost:29092") \
    .option("subscribe", "visits_raw") \
    .option("startingOffsets", "EARLIEST") \
    .format("kafka") \
    .load()


records_to_write = process_input_data(input_data_stream)

write_query = records_to_write.writeStream.format("kafka")\
    .option("checkpointLocation", "/tmp/bde/module5/homework/e2/checkpoint")\
    .options(**get_sink_configuration())\
    .start()

write_query.awaitTermination()

File: 03-stream-processing/solution-proposal/scala/src/main/scala/com/becomedataengineer/exercise2/StreamingJob.scala
package com.becomedataengineer.exercise2

import com.becomedataengineer.exercise2.SinkConfiguration.SinkOptions
import org.apache.spark.sql.SparkSession

object StreamingJob {

  def main(args: Array[String]): Unit = {
    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()
    import sparkSession.implicits._

    val dataFrame = sparkSession.readStream
      .format("kafka")
      .options(Map(
        "kafka.bootstrap.servers" -> "localhost:29092",
        "subscribe" -> "visits_raw",
        "startingOffsets" -> "EARLIEST"
      ))
      .load()

    val recordsToWrite = StreamingJobProcessingLogic.processInputData(dataFrame)

    val writeQuery = recordsToWrite.writeStream
      .option("checkpointLocation", "/tmp/bde/module5/homework/e2/checkpoint")
      .format("kafka").options(SinkOptions)

    writeQuery.start().awaitTermination()
  }

}
